File: vehicle/src/main/java/net/tomofiles/skysign/vehicle/domain/vehicle/Vehicle.java
package net.tomofiles.skysign.vehicle.domain.vehicle;

import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import net.tomofiles.skysign.vehicle.event.EmptyPublisher;
import net.tomofiles.skysign.vehicle.event.Publisher;

@EqualsAndHashCode(of = {"id"})
@ToString
public class Vehicle {
    @Getter
    private final VehicleId id;

    private final Generator generator;
    
    @Getter
    private final boolean isCarbonCopy;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private String vehicleName = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private CommunicationId communicationId = null;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version version;

    @Getter
    @Setter(value = AccessLevel.PACKAGE)
    private Version newVersion;

    @Setter
    private Publisher publisher = new EmptyPublisher();
    
    Vehicle(VehicleId id, boolean isCarbonCopy, Version version, Generator generator) {
        this.id = id;
        this.isCarbonCopy = isCarbonCopy;
        this.version = version;
        this.newVersion = version;

        this.generator = generator;
    }

    static Vehicle newOriginal(VehicleId id, Version version, Generator generator) {
        return new Vehicle(id, false, version, generator);
    } 

    static Vehicle newCarbonCopy(VehicleId id, Version version, Generator generator) {
        return new Vehicle(id, true, version, generator);
    } 

    public void nameVehicle(String name) {
        if (this.isCarbonCopy) {
            throw new CannotChangeVehicleException("cannot change carbon copied vehicle");
        }

        this.vehicleName = name;
        this.newVersion = this.generator.newVersion();
    }

    public void giveCommunication(CommunicationId id) {
        if (this.isCarbonCopy) {
            throw new CannotChangeVehicleException("cannot change carbon copied vehicle");
        }

        if (this.communicationId == null) {
            this.communicationId = id;
            this.newVersion = this.generator.newVersion();
            this.publisher
                    .publish(
                            new CommunicationIdGaveEvent(this.communicationId));
        } else {
            CommunicationId beforeId = this.communicationId;
            this.communicationId = id;
            this.newVersion = this.generator.newVersion();
            this.publisher
                    .publish(
                            new CommunicationIdRemovedEvent(beforeId));
            this.publisher
                    .publish(
                            new CommunicationIdGaveEvent(this.communicationId));
        }
    }

    public void removeCommunication() {
        if (this.isCarbonCopy) {
            throw new CannotChangeVehicleException("cannot change carbon copied vehicle");
        }

        CommunicationId removedId = this.communicationId;
        this.communicationId = null;
        this.newVersion = this.generator.newVersion();
        this.publisher
                .publish(
                        new CommunicationIdRemovedEvent(removedId));
    }
}

File: vehicle/pkg/vehicle/domain/vehicle/vehicle.go
package vehicle

import (
	"errors"
	"vehicle/pkg/vehicle/domain/event"
)

// ID .
type ID string

// CommunicationID .
type CommunicationID string

// Version .
type Version string

const (
	// Original .
	Original = false
	// CarbonCopy .
	CarbonCopy = true
)

var (
	// ErrCannotChange .
	ErrCannotChange = errors.New("cannot change carbon copied vehicle")
)

// Vehicle .
type Vehicle struct {
	id              ID
	name            string
	communicationID CommunicationID
	isCarbonCopy    bool
	version         Version
	newVersion      Version
	gen             Generator
	pub             event.Publisher
}

// SetPublisher .
func (v *Vehicle) SetPublisher(pub event.Publisher) {
	v.pub = pub
}

// GetID .
func (v *Vehicle) GetID() ID {
	return v.id
}

// GetName .
func (v *Vehicle) GetName() string {
	return v.name
}

// GetCommunicationID .
func (v *Vehicle) GetCommunicationID() CommunicationID {
	return v.communicationID
}

// GetVersion .
func (v *Vehicle) GetVersion() Version {
	return v.version
}

// GetNewVersion .
func (v *Vehicle) GetNewVersion() Version {
	return v.newVersion
}

// NameVehicle .
func (v *Vehicle) NameVehicle(name string) error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	v.name = name
	v.newVersion = v.gen.NewVersion()
	return nil
}

// GiveCommunication .
func (v *Vehicle) GiveCommunication(communicationID CommunicationID) error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	if v.pub != nil {
		if v.communicationID == "" {
			v.communicationID = communicationID
			v.newVersion = v.gen.NewVersion()
			v.pub.Publish(CommunicationIdGaveEvent{
				CommunicationID: v.communicationID,
			})
		} else {
			beforeId := v.communicationID
			v.communicationID = communicationID
			v.newVersion = v.gen.NewVersion()
			v.pub.Publish(CommunicationIdRemovedEvent{
				CommunicationID: beforeId,
			})
			v.pub.Publish(CommunicationIdGaveEvent{
				CommunicationID: v.communicationID,
			})
		}
	}
	return nil
}

// RemoveCommunication .
func (v *Vehicle) RemoveCommunication() error {
	if v.isCarbonCopy {
		return ErrCannotChange
	}
	if v.pub != nil {
		v.pub.Publish(CommunicationIdRemovedEvent{
			CommunicationID: v.communicationID,
		})
	}
	v.communicationID = ""
	v.newVersion = v.gen.NewVersion()
	return nil
}

// Generator .
type Generator interface {
	NewID() ID
	NewVersion() Version
}
